bpf/tests/wireguard_native.c

```diff
@@ -0,0 +1,242 @@
+	return TEST_ERROR;
+}
+
+CHECK("tc", "wg_v4_1_missing_dst")
+int wg_v4_1_missing_dst_check(const struct __ctx_buff *ctx)
+{
+	void *data;
+	void *data_end;
+	__u32 *status_code;
+
+	test_init();
+
+	data = (void *)(long)ctx->data;
+	data_end = (void *)(long)ctx->data_end;
+
+	if (data + sizeof(*status_code) > data_end)
+		test_fatal("status code out of bounds");
+
+	status_code = data;
+	assert(*status_code == CTX_ACT_REDIRECT);
+
+	test_finish();
+}
+
+/* Validate the behavior for a terminating source endpoint.
+ *
+ * First we test *with* the sec identity in the mark. This should be
+ * sufficient to trigger encryption, even if the endpoint's ipcache entry
+ * is not available.
+ */
+PKTGEN("tc", "wg_v4_2_src_mark")
+int wg_v4_2_src_mark_pktgen(struct __ctx_buff *ctx)
+{
+	return v4_build_packet(ctx);
+}
+
+SETUP("tc", "wg_v4_2_src_mark")
+int wg_v4_2_src_mark_setup(struct __ctx_buff *ctx)
+{
+	ipcache_v4_add_entry(DST_POD_V4, 0, DST_POD_SEC_IDENTITY,
+			     DST_NODE_V4, ENCRYPT_KEY);
+	ipcache_v4_add_entry_with_mask_size(DST_POD_CIDR_V4, 0, WORLD_ID,
+					    DST_NODE_V4, ENCRYPT_KEY,
+					    v4_pod_cidr_size);
+
+	set_identity_mark(ctx, SRC_POD_SEC_IDENTITY, MARK_MAGIC_IDENTITY);
+
+	/* Jump into the entrypoint */
+	tail_call_static(ctx, entry_call_map, TO_NETDEV);
+	/* Fail if we didn't jump */
+	return TEST_ERROR;
+}
+
+CHECK("tc", "wg_v4_2_src_mark")
+int wg_v4_2_src_mark_check(const struct __ctx_buff *ctx)
+{
+	void *data;
+	void *data_end;
+	__u32 *status_code;
+
+	test_init();
+
+	data = (void *)(long)ctx->data;
+	data_end = (void *)(long)ctx->data_end;
+
+	if (data + sizeof(*status_code) > data_end)
+		test_fatal("status code out of bounds");
+
+	status_code = data;
+	assert(*status_code == CTX_ACT_REDIRECT);
+
+	test_finish();
+}
+
+/* Now test *without* the identity in the mark. This should *not* trigger
+ * encryption:
+ */
+PKTGEN("tc", "wg_v4_3_no_src_mark")
+int wg_v4_3_no_src_mark_pktgen(struct __ctx_buff *ctx)
+{
+	return v4_build_packet(ctx);
+}
+
+SETUP("tc", "wg_v4_3_no_src_mark")
+int wg_v4_3_no_src_mark_setup(struct __ctx_buff *ctx)
+{
+	ipcache_v4_add_entry(DST_POD_V4, 0, DST_POD_SEC_IDENTITY,
+			     DST_NODE_V4, ENCRYPT_KEY);
+	ipcache_v4_add_entry_with_mask_size(DST_POD_CIDR_V4, 0, WORLD_ID,
+					    DST_NODE_V4, ENCRYPT_KEY,
+					    v4_pod_cidr_size);
+
+	/* Jump into the entrypoint */
+	tail_call_static(ctx, entry_call_map, TO_NETDEV);
+	/* Fail if we didn't jump */
+	return TEST_ERROR;
+}
+
+CHECK("tc", "wg_v4_3_no_src_mark")
+int wg_v4_3_no_src_mark_check(const struct __ctx_buff *ctx)
+{
+	void *data;
+	void *data_end;
+	__u32 *status_code;
+
+	test_init();
+
+	data = (void *)(long)ctx->data;
+	data_end = (void *)(long)ctx->data_end;
+
+	if (data + sizeof(*status_code) > data_end)
+		test_fatal("status code out of bounds");
+
+	status_code = data;
+	assert(*status_code == CTX_ACT_OK);
+
+	test_finish();
+}
+
+/* Finally test without the mark, but with the endpoint's ipcache entry: */
```

Comment:
Maybe I'm tripping, but is it worth adding the case in which we don't have endpoint ipcache entry but still local CIDR entry in the ipcache? 🤔

Reply:
Right... I did a quick test and the only CIDR I see from a node are the remote ones 😞 

Don't know if it would make sense to add (i.e., if there's a time window where we would already have CIDRs but not endpoint info locally).

Reply:
I also see the value of having the local PodCIDR in the ipcache entry, yep. It needs some architecture-level thinking though:
* for outbound traffic, we should ideally rely on the source identity in the packet mark. And **not** need an ipcache lookup, or endpoint map lookup.
* for inbound traffic, it would be cool to drop any un-deliverable packet that's intended for the local PodCIDR. And not exit via the default route.

Reply:
yep! And similar in the other inbound programs (from-overlay, from-wireguard, from-container).